include/lbann/execution_contexts/execution_context.hpp

```diff
@@ -0,0 +1,151 @@
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_EXECUTION_CONTEXT_HPP
+#define LBANN_EXECUTION_CONTEXT_HPP
+
+#include "lbann/base.hpp"
+#include "lbann/comm.hpp"
+#include "lbann/io/persist.hpp"
+#include "lbann/utils/threads/thread_pool.hpp"
+#include <cereal/types/utility.hpp>
+
+namespace lbann {
+
+// Forward-declare this.
+class trainer;
+
+struct termination_criteria {
+  El::Int num_steps;
+};
+
+class execution_context {
+public:
+  /** Constructor. */
+  execution_context(observing_ptr<trainer> trainer, lbann_comm *comm, execution_mode mode);
+  /** Destructor. */
+  virtual ~execution_context() = default;
+
+  /** Copy execution_context. */
+  virtual std::unique_ptr<execution_context> copy_execution_context() const {
+    // Use explicit construction of unique pointer since copy
+    // constructor is protected and cannot be accessed in make_unique
+    return std::unique_ptr<execution_context>{new execution_context(*this)};
+  }
+
+  /** Archive for checkpoint and restart */
+  template <class Archive> void serialize( Archive & ar ) {
+    ar(CEREAL_NVP(m_execution_mode),
+       CEREAL_NVP(m_terminate_training),
+       CEREAL_NVP(m_step));
+  }
+
+  /** @brief Current step in the training algorithm
+    *  @detailed Step counts the number of iterations in the training
+    *  algorithm's internal state
+    */
+  El::Int get_step() const noexcept { return m_step; }
+
+  /** @brief Increment the current step in the training algorithm
+    *  @detailed Increment the step count in the training
+    *  algorithm's internal state
+    */
+  void inc_step() noexcept { ++m_step; }
+
+  /** Get the mode that the trainer is currenting executing. */
+  inline void set_execution_mode(execution_mode mode) noexcept {
+    m_execution_mode = mode;
+  }
+
+  /** Get the mode that the trainer is currenting executing. */
+  inline execution_mode get_execution_mode() const noexcept {
+    return m_execution_mode;
+  }
+
+  /** Return true if the flag to stop training is set. */
+  bool get_terminate_training() const {
+    return m_terminate_training;
+  }
+  /** Set the terminate training flag (on or off). */
+  void set_terminate_training(bool f) {
+    m_terminate_training = f;
+  }
+
+  /** Get the execution environment */
+  trainer* get_trainer() {
```

Comment:
Does this need to return a pointer? Unless the trainer can *meaningfully* be null, this should be `trainer&`. Otherwise, `observing_ptr<trainer>` vs the raw pointer notation.